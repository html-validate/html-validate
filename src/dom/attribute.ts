import { Location } from "../context";
import { DynamicValue } from "./dynamic-value";

interface StaticAttribute {
	readonly value: string | null;
}

interface DynamicAttribute {
	readonly value: DynamicValue;
}

/**
 * Type guard for [[Attribute]] testing if the value is static (missing value is
 * considered static as it is a fixed known value).
 *
 * @public
 */
export function isStaticAttribute(attr: Attribute | null): attr is Attribute & StaticAttribute {
	return Boolean(attr && attr.isStatic);
}

/**
 * Type guard for [[Attribute]] testing if the value is dynamic.
 *
 * @public
 */
export function isDynamicAttribute(attr: Attribute | null): attr is Attribute & DynamicAttribute {
	return Boolean(attr && attr.isDynamic);
}

/**
 * DOM Attribute.
 *
 * Represents a HTML attribute. Can contain either a fixed static value or a
 * placeholder for dynamic values (e.g. interpolated).
 */
export class Attribute {
	/** Attribute name */
	public readonly key: string;
	public readonly value: string | DynamicValue | null;
	public readonly keyLocation: Location;
	public readonly valueLocation: Location | null;
	public readonly originalAttribute?: string;

	/**
	 * @param key - Attribute name.
	 * @param value - Attribute value. Set to `null` for boolean attributes.
	 * @param keyLocation - Source location of attribute name.
	 * @param valueLocation - Source location of attribute value.
	 * @param originalAttribute - If this attribute was dynamically added via a
	 * transformation (e.g. vuejs `:id` generating the `id` attribute) this
	 * parameter should be set to the attribute name of the source attribute (`:id`).
	 */
	public constructor(
		key: string,
		value: null | string | DynamicValue | null,
		keyLocation: Location,
		valueLocation: Location | null,
		originalAttribute?: string
	) {
		this.key = key;
		this.value = value;
		this.keyLocation = keyLocation;
		this.valueLocation = valueLocation;
		this.originalAttribute = originalAttribute;

		/* force undefined to null */
		if (typeof this.value === "undefined") {
			this.value = null;
		}
	}

	/**
	 * Flag set to true if the attribute value is static.
	 */
	public get isStatic(): boolean {
		return !this.isDynamic;
	}

	/**
	 * Flag set to true if the attribute value is dynamic.
	 */
	public get isDynamic(): boolean {
		return this.value instanceof DynamicValue;
	}

	/**
	 * Test attribute value.
	 *
	 * @param pattern - Pattern to match value against. RegExp or a string (===)
	 * @param dynamicMatches - If true `DynamicValue` will always match, if false
	 * it never matches.
	 * @returns `true` if attribute value matches pattern.
	 */
	public valueMatches(pattern: RegExp, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: string, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: RegExp | string, dynamicMatches: boolean = true): boolean {
		if (this.value === null) {
			return false;
		}

		/* dynamic values matches everything */
		if (this.value instanceof DynamicValue) {
			return dynamicMatches;
		}

		/* test value against pattern */
		if (pattern instanceof RegExp) {
			return this.value.match(pattern) !== null;
		} else {
			return this.value === pattern;
		}
	}
}
