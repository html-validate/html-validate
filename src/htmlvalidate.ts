import { Config, ConfigLoader } from './config';
import { Engine, TokenDump, EventDump } from './engine';
import { Parser } from './parser';
import { Report } from './reporter';
import { Source } from './context';

class HtmlValidate {
	private globalConfig: Config;

	constructor(options?: any){
		const defaults = Config.empty();
		this.globalConfig = defaults.merge(options ? Config.fromObject(options) : Config.defaultConfig());
	}

	/**
	 * Parse HTML from string.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateString(str: string): Report {
		const source = {
			data: str,
			filename: 'inline',
			line: 1,
			column: 1,
		};
		return this.validateSource(source);
	}

	/**
	 * Parse HTML from source.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateSource(source: Source): Report {
		const config = this.getConfigFor('inline');
		const engine = new Engine(config, Parser);
		return engine.lint([source]);
	}

	/**
	 * Parse HTML from file.
	 *
	 * @param filename {string} - Filename to read and parse.
	 * @return {object} - Report output.
	 */
	public validateFile(filename: string): Report {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.lint(source);
	}

	public dumpTokens(filename: string): TokenDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTokens(source);
	}

	public dumpEvents(filename: string): EventDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpEvents(source);
	}

	public dumpTree(filename: string): string[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTree(source);
	}

	/**
	 * Get parser for given filename.
	 */
	public getParserFor(source: Source): Parser {
		const config = this.getConfigFor(source.filename);
		return new Parser(config);
	}

	/**
	 * Get configuration for given filename.
	 */
	public getConfigFor(filename: string): Config {
		const loader = new ConfigLoader();
		const config = loader.fromTarget(filename);
		const merged = this.globalConfig.merge(config);
		merged.init();
		return merged;
	}
}

export default HtmlValidate;
