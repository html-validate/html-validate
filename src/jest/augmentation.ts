/** module augmentation: jest.d.ts */
import { type ConfigData } from "../config";
import { type Message } from "../message";

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace -- module augmentation */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/no-empty-object-type, @typescript-eslint/no-unused-vars -- to match jest declaration */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toHaveError(error: Partial<Message>): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(errors: Array<[string, string] | Record<string, unknown>>): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 * @param filename - Optional filename used when matching transformer and
			 * loading configuration.
			 */
			toHTMLValidate(filename?: string): R;
			toHTMLValidate(config: ConfigData, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, config: ConfigData, filename?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with snapshot.
			 */
			toMatchCodeframe(snapshot?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with inline snapshot.
			 */
			toMatchInlineCodeframe(snapshot?: string): R;
		}
	}
}
/** module augmentation end */
