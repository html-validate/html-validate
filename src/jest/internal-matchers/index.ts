export { toBeToken } from "./to-be-token";
