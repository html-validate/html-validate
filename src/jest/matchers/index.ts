export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
