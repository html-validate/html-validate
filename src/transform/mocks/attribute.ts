import { DynamicValue } from "../../dom";
import { type AttributeData } from "../../parser";

export function* processAttribute(attr: AttributeData): Iterable<AttributeData> {
	/* handle foo="{{ bar }}" as "foo" with a dynamic value (interpolated) */
	if (typeof attr.value === "string" && /{{.*}}/.exec(attr.value)) {
		yield {
			...attr,
			value: new DynamicValue(attr.value),
		};
		return;
	}

	/* passthru original attribute */
	yield attr;

	/* handle "dynamic-foo" as alias for "foo" with dynamic value */
	if (attr.key.startsWith("dynamic-")) {
		yield {
			...attr,
			key: attr.key.replace("dynamic-", ""),
			value: new DynamicValue(attr.value as string),
			originalAttribute: attr.key,
		};
	}
}
