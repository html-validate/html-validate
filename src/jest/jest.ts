import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
