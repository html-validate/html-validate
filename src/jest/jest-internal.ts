import "./jest";

import { Location } from "../context";
import { TokenType } from "../lexer";
import { toBeToken } from "./internal-matchers";

interface TokenMatcher {
	type: TokenType;
	location?: Partial<Location>;
	data?: any;
}

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-unused-vars */
		interface Matchers<R, T = {}> {
			toBeToken(expected: TokenMatcher): R;
		}
	}
}

expect.extend({
	toBeToken,
});
