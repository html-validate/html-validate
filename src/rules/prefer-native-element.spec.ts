import { HtmlValidate } from "../htmlvalidate";
import "../jest";
import { processAttribute } from "../transform/mocks/attribute";

describe("rule prefer-native-element", () => {
	describe("default config", () => {
		let htmlvalidate: HtmlValidate;

		beforeAll(() => {
			htmlvalidate = new HtmlValidate({
				root: true,
				rules: { "prefer-native-element": "error" },
			});
		});

		it("should not report error when using role without native equivalent element", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<div role="unknown"></div>');
			expect(report).toBeValid();
		});

		it("should not report error when role is boolean", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString("<div role></div>");
			expect(report).toBeValid();
		});

		it("should not report error for dynamic attributes", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<input dynamic-role="main">', {
				processAttribute,
			});
			expect(report).toBeValid();
		});

		it("should not report error when element has redundant role", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<main role="main"></main>');
			expect(report).toBeValid();
		});

		it("should report error when using role with native equivalent element", () => {
			expect.assertions(2);
			const report = htmlvalidate.validateString('<div role="main"></div>');
			expect(report).toBeInvalid();
			expect(report).toHaveError(
				"prefer-native-element",
				"Prefer to use the native <main> element",
			);
		});

		it("should handle unquoted role", () => {
			expect.assertions(2);
			const report = htmlvalidate.validateString("<div role=main></div>");
			expect(report).toBeInvalid();
			expect(report).toHaveError(
				"prefer-native-element",
				"Prefer to use the native <main> element",
			);
		});

		it("smoketest", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateFile("test-files/rules/prefer-native-element.html");
			expect(report).toMatchCodeframe();
		});
	});

	it("should not report error when role is excluded", () => {
		expect.assertions(2);
		const htmlvalidate = new HtmlValidate({
			root: true,
			rules: { "prefer-native-element": ["error", { exclude: ["main"] }] },
		});
		const valid = htmlvalidate.validateString('<div role="main"></div>');
		const invalid = htmlvalidate.validateString('<div role="article"></div>');
		expect(valid).toBeValid();
		expect(invalid).toBeInvalid();
	});

	it("should report error only for included roles", () => {
		expect.assertions(2);
		const htmlvalidate = new HtmlValidate({
			root: true,
			rules: { "prefer-native-element": ["error", { include: ["main"] }] },
		});
		const valid = htmlvalidate.validateString('<div role="article"></div>');
		const invalid = htmlvalidate.validateString('<div role="main"></div>');
		expect(valid).toBeValid();
		expect(invalid).toBeInvalid();
	});

	it("should contain documentation", async () => {
		expect.assertions(1);
		const htmlvalidate = new HtmlValidate({
			root: true,
			rules: { "prefer-native-element": "error" },
		});
		const context = {
			role: "the-role",
			replacement: "the-replacement",
		};
		const docs = await htmlvalidate.getContextualDocumentation({
			ruleId: "prefer-native-element",
			context,
		});
		expect(docs).toMatchInlineSnapshot(`
			{
			  "description": "Instead of using the WAI-ARIA role "the-role" prefer to use the native <the-replacement> element.",
			  "url": "https://html-validate.org/rules/prefer-native-element.html",
			}
		`);
	});
});
