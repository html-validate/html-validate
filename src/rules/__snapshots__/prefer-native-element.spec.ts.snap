// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule prefer-native-element default config smoketest 1`] = `
Array [
  Object {
    "errorCount": 22,
    "filePath": "test-files/rules/prefer-native-element.html",
    "messages": Array [
      Object {
        "column": 6,
        "context": Object {
          "replacement": "article",
          "role": "article",
        },
        "line": 3,
        "message": "Prefer to use the native <article> element",
        "offset": 33,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 14,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "header",
          "role": "banner",
        },
        "line": 4,
        "message": "Prefer to use the native <header> element",
        "offset": 60,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 13,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "button",
          "role": "button",
        },
        "line": 5,
        "message": "Prefer to use the native <button> element",
        "offset": 86,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 13,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "td",
          "role": "cell",
        },
        "line": 6,
        "message": "Prefer to use the native <td> element",
        "offset": 112,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 11,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "input",
          "role": "checkbox",
        },
        "line": 7,
        "message": "Prefer to use the native <input> element",
        "offset": 136,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 15,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "aside",
          "role": "complementary",
        },
        "line": 8,
        "message": "Prefer to use the native <aside> element",
        "offset": 164,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 20,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "footer",
          "role": "contentinfo",
        },
        "line": 9,
        "message": "Prefer to use the native <footer> element",
        "offset": 197,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 18,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "figure",
          "role": "figure",
        },
        "line": 10,
        "message": "Prefer to use the native <figure> element",
        "offset": 228,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 13,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "form",
          "role": "form",
        },
        "line": 11,
        "message": "Prefer to use the native <form> element",
        "offset": 254,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 11,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "hN",
          "role": "heading",
        },
        "line": 12,
        "message": "Prefer to use the native <hN> element",
        "offset": 278,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 14,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "input",
          "role": "input",
        },
        "line": 13,
        "message": "Prefer to use the native <input> element",
        "offset": 305,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 12,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "a",
          "role": "link",
        },
        "line": 14,
        "message": "Prefer to use the native <a> element",
        "offset": 330,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 11,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "ul",
          "role": "list",
        },
        "line": 15,
        "message": "Prefer to use the native <ul> element",
        "offset": 354,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 11,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "select",
          "role": "listbox",
        },
        "line": 16,
        "message": "Prefer to use the native <select> element",
        "offset": 378,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 14,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "li",
          "role": "listitem",
        },
        "line": 17,
        "message": "Prefer to use the native <li> element",
        "offset": 405,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 15,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "main",
          "role": "main",
        },
        "line": 18,
        "message": "Prefer to use the native <main> element",
        "offset": 433,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 11,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "nav",
          "role": "navigation",
        },
        "line": 19,
        "message": "Prefer to use the native <nav> element",
        "offset": 457,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 17,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "progress",
          "role": "progressbar",
        },
        "line": 20,
        "message": "Prefer to use the native <progress> element",
        "offset": 487,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 18,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "input",
          "role": "radio",
        },
        "line": 21,
        "message": "Prefer to use the native <input> element",
        "offset": 518,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 12,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "section",
          "role": "region",
        },
        "line": 22,
        "message": "Prefer to use the native <section> element",
        "offset": 543,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 13,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "table",
          "role": "table",
        },
        "line": 23,
        "message": "Prefer to use the native <table> element",
        "offset": 569,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 12,
      },
      Object {
        "column": 6,
        "context": Object {
          "replacement": "textarea",
          "role": "textbox",
        },
        "line": 24,
        "message": "Prefer to use the native <textarea> element",
        "offset": 594,
        "ruleId": "prefer-native-element",
        "severity": 2,
        "size": 14,
      },
    ],
    "source": "<div role=\\"unknown\\"></div>

<div role=\\"article\\"></div>
<div role=\\"banner\\"></div>
<div role=\\"button\\"></div>
<div role=\\"cell\\"></div>
<div role=\\"checkbox\\"></div>
<div role=\\"complementary\\"></div>
<div role=\\"contentinfo\\"></div>
<div role=\\"figure\\"></div>
<div role=\\"form\\"></div>
<div role=\\"heading\\"></div>
<div role=\\"input\\"></div>
<div role=\\"link\\"></div>
<div role=\\"list\\"></div>
<div role=\\"listbox\\"></div>
<div role=\\"listitem\\"></div>
<div role=\\"main\\"></div>
<div role=\\"navigation\\"></div>
<div role=\\"progressbar\\"></div>
<div role=\\"radio\\"></div>
<div role=\\"region\\"></div>
<div role=\\"table\\"></div>
<div role=\\"textbox\\"></div>
",
    "warningCount": 0,
  },
]
`;

exports[`rule prefer-native-element should contain contextual documentation 1`] = `
Object {
  "description": "Instead of using the WAI-ARIA role \\"the-role\\" prefer to use the native <the-replacement> element.",
  "url": "https://html-validate.org/rules/prefer-native-element.html",
}
`;

exports[`rule prefer-native-element should contain documentation 1`] = `
Object {
  "description": "Instead of using WAI-ARIA roles prefer to use the native HTML elements.",
  "url": "https://html-validate.org/rules/prefer-native-element.html",
}
`;
